docs/developer-docs/security/security-best-practices/dos.mdx

```diff
@@ -4,32 +4,72 @@ keywords: [security, concept, denial of service, dos, cycles, captcha]
+### Understanding DoS and DDoS Attacks
 
-* For query calls that cause significant computation and don't modify the state, it is advisable to not execute the expensive computation if the method is called as an update. However, keep in mind that query calls [don't provide authenticity guarantees](/docs/current/developer-docs/security/security-best-practices/data-integrity-and-authenticity#certify-query-responses-if-they-are-relevant-for-security), so this is a trade-off. Unfortunately, the execution mode of the query (whether it was called as query or update) is currently not directly exposed to the user code. However, one can call a method such as `ic0.data_certificate_present()` which returns `1` when called as query, and `0` for update methods. See the Interface Specification [section on certified data](/docs/current/references/ic-interface-spec#system-api-certified-data).
+A Denial of Service (DoS) attack aims to make a system unavailable by overwhelming it with requests or data. A Distributed Denial of Service (DDoS) attack is a more sophisticated version, where the attack originates from multiple sources, making it harder to block. 
 
-* Expensive calls that only need to be called from other canisters can require some amount of cycles to be sent along with the call to compensate for the cycles consumed by the execution.
+In the case of canisters, such attacks can quickly deplete cycles, making the canister unable to process legitimate requests. Learn more about DoS in the [Security Best Practices: Denial of Service](https://internetcomputer.org/docs/current/developer-docs/security/security-best-practices/dos) guide.
 
-* Finally, it is also an option to charge for ingress messages, but that is not currently supported by the protocol itself and a custom solution would need to be designed.
+### Preventing DoS and DDoS Attacks
 
-## For expensive calls, consider using captchas or proof of work
+To protect your canisters from DoS and DDoS attacks, consider the following strategies:
+- **Bot prevention techniques**: Use methods like captchas or proof of work to ensure only legitimate users can access your canister. Captchas help verify that the user is human, while proof of work requires the user to spend computational resources to proceed, deterring automated attacks. An example implementation can be found in the [Internet Identity project](https://github.com/dfinity/internet-identity).
+- **Monitor cycles usage**: Regularly track your canisters cycles consumption and set alerts for any sudden spikes that may indicate an attack.
+- **Ingress message charging**: While charging for ingress messages (external requests to the canister) is not natively supported, custom solutions could be implemented to make sure that any expensive actions have costs associated with them.
```

Comment:
What would that look like?
Requiring an approval on a ledger where the dapp can charge the user from?

Reply:
It's similar the wallet (proxy) canister idea we use with dfx.

Reply:
It's similar the wallet (proxy) canister idea we use with dfx.

Reply:
I'm not 100% sure I understand what you mean. Would you mind writing a suggested recommendation in this PR?